tests/test_project.py

```diff
@@ -540,6 +546,24 @@ def test_run_with_selection(self):
             else:
                 self.assertFalse(job.isfile('world.txt'))
 
+    def test_run_with_operation_selection(self):
+        project = self.mock_project()
+        even_jobs = [job for job in project if job.sp.b % 2 == 0]
+        with add_cwd_to_environment_pythonpath():
+            with switch_to_directory(project.root_directory()):
+                with self.assertRaises(ValueError):
+                    # The names argument must be a sequence of strings, not a string.
+                    project.run(names='op1')
+                project.run(names=['non-existent-op'])
+                self.assertFalse(any(job.isfile('world.txt') for job in even_jobs))
+                self.assertFalse(any(job.doc.get('test') for job in project))
+                project.run(names=['op1', 'non-existent-op'])
+                self.assertTrue(all(job.isfile('world.txt') for job in even_jobs))
+                self.assertFalse(any(job.doc.get('test') for job in project))
+                project.run(names=['op[^3]', 'non-existent-op'])
```

Comment:
Can you also add an assertion that `dynamic` is not in the doc (to ensure `op3` didn't execute)?

Reply:
Done.